micropython/umqtt.robust/umqtt/robust.py

```diff
@@ -42,3 +42,12 @@ def wait_msg(self):
             except OSError as e:
                 self.log(False, e)
             self.reconnect()
+
+    def check_msg(self):
+        while 1:
+            self.sock.setblocking(False)
+            try:
+                return super().wait_msg()
+            except OSError as e:
+                self.log(False, e)
+            self.reconnect()
```

Comment:
This could be updated to add a new optional arg to `def wait_msg(self, blocking=True)` and if blocking is False, then
 ```
except OSError as e:
    if not blocking:
        raise
```
which would mean the error would be propogated back up to `check_msg()` to handle.

```suggestion
    def check_msg(self):
        self.sock.setblocking(False)
        try:
            return super().wait_msg()
        except OSError as e:
            self.log(False, e)
            self.reconnect()
        return None
```
I also don't think check_msg() should be a loop, there's nothing wrong with checking once (in non blocking state), still reconnect on error, but just exit straight away then. This still preserves the intention of the functions I think?

Reply:
I don't think there's a **right** answer to this, it depends to some degree on the application.

When calling a **robust** `check_msg()`, my assumption would have been that I can rely on this (more so than umqtt.simple) to deliver me a message if one exists - I believe this is the original author's intention too since `check_msg()` was not overridden in robust and calls back to the overridden `wait_msg()` which does try to be more reliable. I'm convinced `check_msg()` should loop (bear in mind, if there's an exception in `reconnect()`, it will be raised, so the only occasion where we don't return reasonably quickly to `check_msg()`'s caller is if `reconnect()` consistently works and `super().wait_msg()` consistently fails - a highly unusual situation I'd say).

To help us reach consensus, is it worth considering two applications:
1. **`check_msg()` is periodically called on a main loop or from within an async co-routine**
 In this case, I think your suggestion would work reasonably well; in the event of a `reconnect()`, messages would be delayed only by the length of a loop. My suggested implementation does spend time trying a little harder, but delivers a message ASAP.
2. **`check_msg()` is called once having solicited a response from other nodes connected to the MQTT broker (possibly after a short wait for responses)**
 In this case, returning from `check_msg()` having knowingly failed seems inappropriate to me and would need to be clearly stated in its usage documentation "check_msg() should be called at least twice (in what hard-to-detect circumstances?)" Again, my implementation does spend time trying a little harder, but delivers a message first time around if available.

Perhaps it is this specific point that is most concerning, "**spend time** trying a little harder". That's fair critique if returning quickly takes precedence over being robust for `check_msg()`. That said, if that's the route we go down, I believe we need to signal the failure in some way, possibly by returning `False` instead of `None` on your last line.

Reply:
Yes you make some very good points in that there's no one definitive right answer.

Looking through the suggestions, I now feel like check_msg should retry just once if needed, ie call wait_msg, on exception call reconnect then call wait_msg again. 

If it fails again it should still just return as it means the server is probably down. 

I definitely believe check_msg should not continue to loop on failure because that will be blocking and prevent the caller from logging, notifying user of failure etc, basically the same as a wait_msg if the server is down.